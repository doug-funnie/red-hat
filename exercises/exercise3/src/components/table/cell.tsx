import glamorous, { CSSProperties } from 'glamorous';
import { ThemedProps } from 'styles/theme';

type Props = ThemedProps<{
  grow?: number;
  isHeading?: boolean;
  hideOnMobile?: boolean;
}>;

const defaultProps: Partial<Props> = {
  grow: 1,
  isHeading: false,
  hideOnMobile: false,
};

export const Cell = glamorous.div<Props>(({
  theme,
  grow,
  hideOnMobile,
  isHeading,
}) => {
  const styles: CSSProperties = {
    color: theme.colors.textLight,
    flex: `${grow} 0 100px`,
    fontSize: 14,
    fontWeight: isHeading ? 500 : 400,
    overflow: 'hidden',
    padding: 10,
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  };

  if (hideOnMobile) {
    styles.display = 'none';
    styles[theme.breakpoints.minWidthSM] = {
      display: 'block',
    };
  }

  return styles;
});

Cell.defaultProps = defaultProps;
