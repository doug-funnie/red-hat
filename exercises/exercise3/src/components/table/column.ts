export type CellRenderer<D> = (props: {
  data: D;
  index: number;
  column: Column<D>;
}) => React.ReactNode;

export type LabelRenderer<D> = (props: { column: Column<D>, index: number }) => React.ReactNode;

export type KeyExtractor<D> = (props: { data: D, index: number }) => string | number;

export type RowClickHandler<D> = (props: { data: D, index: number, event: React.MouseEvent<HTMLDivElement> }) => void;

export interface Column<D> {
  id: string | number;
  renderCell: CellRenderer<D>;
  renderLabel: LabelRenderer<D>;
  grow?: number;
  hideOnMobile?: boolean;
}
