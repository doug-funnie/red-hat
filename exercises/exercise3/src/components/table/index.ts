export {
  Table,
  KeyExtractor,
} from './table';
export {
  Column,
  CellRenderer,
  LabelRenderer,
} from './column';
export { EntryClickHandler } from './entry';
