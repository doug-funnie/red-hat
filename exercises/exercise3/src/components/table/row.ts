import glamorous, { CSSProperties } from 'glamorous';
import { rgba } from 'polished';
import { ThemedProps } from 'styles/theme';

type Props = ThemedProps<{
  isHeading?: boolean;
  onClick?(evt: React.MouseEvent<HTMLDivElement>): void;
}>;

const defaultProps: Partial<Props> = {
  isHeading: false,
};

export const Row = glamorous.div<Props>(({
  theme,
  isHeading,
  onClick,
}) => {
  const styles: CSSProperties = {
    alignItems: 'center',
    borderRadius: 4,
    display: 'flex',
    marginBottom: 5,
    transition: 'background-color 150ms ease-in-out',
    backgroundColor: !isHeading && theme.colors.white,
  };

  if (onClick) {
    return {
      cursor: onClick && 'pointer',
      ...styles,
      ':hover': {
        backgroundColor: rgba(theme.colors.red, 0.2),
      },
    };
  }
  return styles;
});

Row.defaultProps = defaultProps;
