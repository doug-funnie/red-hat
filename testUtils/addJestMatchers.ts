// import serializer from 'jest-glamor-react';
import { toMatchSnapshot } from 'jest-snapshot';

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeCalledWithSnapshot(snapshotName?: string): void;
    }
  }
}

// expect.addSnapshotSerializer(serializer);

expect.extend({
  toBeCalledWithSnapshot(received, snapshotName) {
    if (received.mock.calls.length === 0) {
      return {
        message: () => 'Mock function was not called',
        pass: false,
      };
    }

    return toMatchSnapshot.call(this, received.mock.calls, snapshotName);
  },
});
