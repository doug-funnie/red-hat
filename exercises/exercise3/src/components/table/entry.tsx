import * as React from 'react';
import { Cell } from './cell';
import { Column } from './column';
import { Row } from './row';

export type EntryClickHandler<D> = (props: { data: D, index: number, event: React.MouseEvent<HTMLDivElement> }) => void;

export interface Props<D = any> {
  data: D;
  index: number;
  columns: Column<D>[];
  onClick: EntryClickHandler<D>;
}

export class Entry extends React.Component<Props> {
  private handleRowClick = (event: React.MouseEvent<HTMLDivElement>) => {
   this.props.onClick({
      event,
      data: this.props.data,
      index: this.props.index,
    });
  }

  public render() {
    const {
      data,
      index,
      columns,
      onClick,
    } = this.props;

    return (
      <Row onClick={onClick ? this.handleRowClick : undefined}>
        {columns.map((column) => (
          <Cell key={column.id} grow={column.grow} hideOnMobile={column.hideOnMobile}>
            {column.renderCell({ data, index , column })}
          </Cell>
        ))}
      </Row>
    );
  }
}
