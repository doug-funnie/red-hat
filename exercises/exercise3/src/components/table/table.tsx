import * as React from 'react';
import { Cell } from './cell';
import { Column } from './column';
import { Entry, EntryClickHandler } from './entry';
import { Row } from './row';

export type KeyExtractor<D> = (props: { data: D, index: number }) => string | number;

export interface Props<D> {
  data: D[];
  columns: Column<D>[];
  keyExtractor: KeyExtractor<D>;
  onEntryClick?: EntryClickHandler<D>;
}

export function Table<D>(props: Props<D>) {
  return (
    <div>
      <Row isHeading>
        {props.columns.map((c, i) => (
          <Cell
            isHeading
            key={c.id}
            grow={c.grow}
            hideOnMobile={c.hideOnMobile}
          >
            {c.renderLabel({ column: c, index: i })}
          </Cell>
        ))}
      </Row>
      {props.data.map((d, i) => (
        <Entry
          key={props.keyExtractor({ data: d, index: i })}
          data={d}
          index={i}
          columns={props.columns}
          onClick={props.onEntryClick}
        />
      ))}
    </div>
  );
}
