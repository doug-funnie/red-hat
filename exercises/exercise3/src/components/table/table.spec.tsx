import { shallow } from 'enzyme';
import * as React from 'react';
import '../../../../../testUtils/addJestMatchers';
import { Entry } from './entry';
import { Props, Table } from './table';

const props: Props<number> = {
  data: [1],
  keyExtractor: jest.fn((v) => v),
  onEntryClick: jest.fn(),
  columns: [{
    id: 1,
    renderLabel: jest.fn(),
    renderCell: jest.fn(),
  }],
};

it('calls keyExtractor', () => {
  shallow(<Table {...props} />);
  expect(props.keyExtractor).toBeCalledWithSnapshot();
});

it('calls renderLabel', () => {
  shallow(<Table {...props} />);
  expect(props.columns[0].renderLabel).toBeCalledWithSnapshot();
});

it('passes onEntry click to the entries', () => {
  const view = shallow(<Table {...props} />);
  expect(view.find(Entry).props().onClick).toBe(props.onEntryClick);
});
