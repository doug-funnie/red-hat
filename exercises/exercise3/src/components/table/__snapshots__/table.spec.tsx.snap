// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls keyExtractor 1`] = `
Array [
  Array [
    Object {
      "data": 1,
      "index": 0,
    },
  ],
]
`;

exports[`calls renderLabel 1`] = `
Array [
  Array [
    Object {
      "column": Object {
        "id": 1,
        "renderCell": [MockFunction],
        "renderLabel": [MockFunction] {
          "calls": [Circular],
        },
      },
      "index": 0,
    },
  ],
]
`;
